### PR TITLE
[ASP.NET Core] Add unit test for validating diagnostic source event subscription

### PR DESCRIPTION
**Pull request by vishweshbankwar**
Related Issue #3495 

Related PR #3519

### CODE REVIEW

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs [hunk 1]
```diff
@@ -694,6 +695,136 @@ public async Task ShouldExportActivityWithOneOrMoreExceptionFilters(int mode)
...
+                {
+                    // ignore exception
+                }
+            }
+
+            Assert.Equal(1, numberOfExceptionCallbacks);
+        }
+
+        [Fact(Skip = "Pending Changes https://github.com/open-telemetry/opentelemetry-dotnet/issues/3495")]
+        public async Task DiagnosticSourceExceptionCallBackIsNotReceivedForExceptionsHandledInMiddleware()
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs:776]: Trying this pattern as it is much simpler to set up.

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs [hunk 2]
```diff
@@ -694,6 +695,136 @@ public async Task ShouldExportActivityWithOneOrMoreExceptionFilters(int mode)
...
+                    new TestHttpInListener(new AspNetCoreInstrumentationOptions())
+                    {
+                        OnExceptionCallback = (activity, payload) =>
+                        {
+                            numberOfExceptionCallbacks++;
+                        },
+                    }))
+                    .Build();
+
+            var builder = WebApplication.CreateBuilder();
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs:792]: Trying this pattern as it is much simpler to set up.

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs [hunk 3]
```diff
@@ -694,6 +695,136 @@ public async Task ShouldExportActivityWithOneOrMoreExceptionFilters(int mode)
...
+            void ConfigureTestServices(IServiceCollection services)
+            {
+                this.tracerProvider = Sdk.CreateTracerProviderBuilder()
+                    .AddAspNetCoreInstrumentation(new AspNetCoreInstrumentation(
+                        new TestHttpInListener(new AspNetCoreInstrumentationOptions())
+                        {
+                            OnExceptionCallback = (activity, payload) =>
+                            {
+                                numberOfExceptionCallbacks++;
+                            },
```
Comment (alanwest) [inline test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs:748]: Not a blocker for this PR, but...

Just like OnCustom, ideally we'd capture the event name for OnException and assert we got the expected one.  But we don't have it.

Seems the AspNetCore and Http instrumentation are the only ones that use OnException. Maybe we could migrate these usages to OnCustom in another PR. Other option I guess would be to pass the event name to OnException.

**Review by alanwest — APPROVED** [review]
LGTM.. just a thought for a potential follow on

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs (hunk 2 above)
Comment (alanwest) [reply test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs:792]: I like it.

Comment (cijothomas) [reply test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs:792]: https://github.com/open-telemetry/opentelemetry-dotnet/pull/3394/files

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs (hunk 3 above)
Comment (vishweshbankwar) [reply test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs:748]: Yes - I was thinking to get rid of OnStart, OnStop and OnException as per discussion [here](https://github.com/open-telemetry/opentelemetry-dotnet/pull/3519#discussion_r936091699). Wanted to ask you and @cijothomas if you see any issues if we do that?.

Comment (alanwest) [reply test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs:748]: I'm in full support of ripping this bandaid off 🩹 sooner than later. This PR just made it look like you were considering doing things incrementally.

Comment (vishweshbankwar) [reply test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs:748]: Yes - wanted to keep that change separate as that will involve SDK changes.

**Review by cijothomas — APPROVED** [review]

**Pull request merged**